### DISCUSSION TITLE
Nitrokey HSM 2 (SmartCard-HSM) and Public Key Auth

### DISCUSSION THREAD

**Discussion by Viajaz** [OpenSC/OpenSC#3206] [Q&A] [answered]
#594 was addressed with #1711 but there is no documentation on how to properly use the public key authentication feature (#3205) so I've been experimenting with four Nitrokey HSM 2 devices (CA with three key custodians).

I initialised the specific Nitrokey HSM 2 device that is to require public key auth:
```bash
sc-hsm-tool --create-dkek-share test-ca1-dkek.pbe --pwd-shares-threshold 2 --pwd-shares-total 3
sc-hsm-tool --initialize --dkek-shares 1 --public-key-auth 3 --required-pub-keys 2 --label "TEST-CA1"
sc-hsm-tool --import-dkek-share test-ca1-dkek.pbe --pwd-shares-total 3
sc-hsm-tool --register-public-key keycustodian_1.asn1
sc-hsm-tool --register-public-key keycustodian_2.asn1
sc-hsm-tool --register-public-key keycustodian_3.asn1
```

`sc-hsm-tool --public-key-auth-status` results in:
```
Using reader with a card: Nitrokey Nitrokey HSM 0
Number of public keys:     3
Missing public keys:       0
Required pubkeys for auth: 2
Authenticated public keys: 0
```

However, I am still able to use `opensc-pkcs11` with user PIN to create and use key material, I thought intialising the device in this mode is suppose to replace (block) that authentication method? Did I do something wrong or is there a problem? How do I use this feature properly?
@CardContact

**Comment by CardContact** [comment]
There is a [presentation](www.smartcard-hsm.com/docs/SmartCard-HSM_n-of-m_Authentication_V1.0_2015-03-25.pdf) and a [HowTo](n-of-m_Authentication_HowTo_V1_2015-09-07.pdf) that explains how n-of-m public key authentication works.

Currently there is only initialization support in sc-hsm-tool, which was contributed by a user. There no support in OpenSC to actually perform the authentication step.

To use n-of-m public key authentication, you will need to use the Java API or the [PKI-as-a-Service](https://github.com/CardContact/deploy-pki-as-a-service).

There are ideas how to integrate n-of-m public key authentication as discussed in #594, but there was very little demand so far.

**Reply by Viajaz** [reply]
Why does it still let me use PIN even after initiation though?

**Reply by CardContact** [reply, marked as answer]
Because you can configure "PKA Only", "PKA and PIN" or "PKA or PIN". Don't know which option sc-hsm-tool sets.

**Reply by Viajaz** [reply]
I'm moving my discussion to https://cdn.cardcontact.de/boards/1/topics/34 as it appears it isn't relevant to OpenSC.
